virtualizarr/writers/icechunk.py

```diff
@@ -112,58 +124,145 @@ def write_variable_to_icechunk(
         )
 
 
+def num_chunks(
+    array,
+    axis: int,
+):
+    return array.shape[axis] // array.chunks[axis]
+
+
+def resize_array(
+    group: "Group",
+    name: str,
+    var: Variable,
+    append_axis: int,
+):  # -> "Array":
+    existing_array = group[name]
+    new_shape = list(existing_array.shape)
+    new_shape[append_axis] += var.shape[append_axis]
+    return existing_array.resize(tuple(new_shape))
+
+
+def get_axis(
+    dims: list[str],
+    dim_name: str,
+) -> int:
+    return dims.index(dim_name)
+
+
+import zarr
+
+
+def _check_compatibile_arrays(
+    ma: ManifestArray, existing_array: zarr.core.array.Array, append_axis: int
+):
+    manifest_api._check_same_dtypes([ma.dtype, existing_array.dtype])
+    # this is kind of gross - _v3_codec_pipeline returns a tuple
+    # Question: Does anything need to be done to apply the codecs to the new manifest array?
+    manifest_api._check_same_codecs(
+        [list(ma.zarray._v3_codec_pipeline()), existing_array.metadata.codecs]
+    )
+    manifest_api._check_same_chunk_shapes([ma.chunks, existing_array.chunks])
+    manifest_api._check_same_ndims([ma.ndim, existing_array.ndim])
+    arr_shapes = [ma.shape, existing_array.shape]
+    manifest_api._check_same_shapes_except_on_concat_axis(arr_shapes, append_axis)
+
+
 def write_virtual_variable_to_icechunk(
     store: "IcechunkStore",
     group: "Group",
     name: str,
     var: Variable,
+    append_dim: Optional[str] = None,
 ) -> None:
     """Write a single virtual variable into an icechunk store"""
     ma = cast(ManifestArray, var.data)
     zarray = ma.zarray
+    mode = store.mode.str
+
+    dims = var.dims
+    append_axis, existing_num_chunks, arr = None, None, None
+    if mode == "a" and append_dim in dims:
+        existing_array = group[name]
+        append_axis = get_axis(dims, append_dim)
+        # check if arrays can be concatenated
+        _check_compatibile_arrays(ma, existing_array, append_axis)
+
+        # determine number of existing chunks along the append axis
+        existing_num_chunks = num_chunks(
+            array=group[name],
+            axis=append_axis,
+        )
 
-    # creates array if it doesn't already exist
-    arr = group.require_array(
-        name=name,
-        shape=zarray.shape,
-        chunk_shape=zarray.chunks,
-        dtype=encode_dtype(zarray.dtype),
-        codecs=zarray._v3_codec_pipeline(),
-        dimension_names=var.dims,
-        fill_value=zarray.fill_value,
-        # TODO fill_value?
-    )
-
-    # TODO it would be nice if we could assign directly to the .attrs property
-    for k, v in var.attrs.items():
-        arr.attrs[k] = encode_zarr_attr_value(v)
-    arr.attrs["_ARRAY_DIMENSIONS"] = encode_zarr_attr_value(var.dims)
+        # resize the array
+        arr = resize_array(
+            group=group,
+            name=name,
+            var=var,
+            append_axis=append_axis,
+        )
```

Comment:
Here you determine `existing_num_chunks`, but then don't actually use it until inside `write_manifest_virtual_refs`. I think you could move the `num_chunks` call inside `write_manifest_virtual_refs`, and eliminate the need to pass the `existing_num_chunks` arg down.

Reply:
Ah yes you're right but the challenge is, after we resize the array, than the function existing_num_chunks will not return the right size. I will think about if there is a better way to handle this, so we don't have to pass the existing_num_chunks arg around